annotations showing in itemTree

abaevbog opened the pull request:
- Annotations are displayed in itemTree under their file attachments on a third level. The annotation spans the entire row.
  - Search matches the actual attachment instead of it's parent file.
  - Can create child notes from annotations of the same item or standalone notes from annotations across different items from the context menu.

larmarange commented:
Such functionality would be of great value for QDA using zotero. Is it planned for version 7?

Thanks for everything

abaevbog commented:
@dstillman, so this is still in progress and quite a few things need tweaking - but the general functionality is working after rebase. 

The main question I have now with the new design is: what do we show in the itemPane when an annotation is selected? There is no "title" per se, so there is nothing to put into the title field at the top. The only collapsible section that seems appropriate is the section with the preview of that single annotation. But it feels like there would be a lot of empty space then...

dstillman commented:
> The only collapsible section that seems appropriate is the section with the preview of that single annotation. But it feels like there would be a lot of empty space then...

I think that's OK. First, they can show as fully expanded, unlike the previews in the reader sidebar. Images annotations can show pretty large. And if you select multiple, you'll see them all. Showing these in the items list is largely to facilitate multi-selection via search.

For multiple, could also show a button for Create Note from Annotations so it's not just in the context menu.

There might be other possibilities down the line, like showing the notes where you've added the annotations or things like that.

adomasven commented on chrome/content/zotero/itemTree.jsx:
```diff
@@ -1899,8 +1925,9 @@ var ItemTree = class ItemTree extends LibraryTree {
-		return this.getRow(index).ref.isRegularItem();
-	};
+		let item = this.getRow(index).ref;
+		return item.isRegularItem() || (item.isFileAttachment() && item.numAnnotations());
```
I don't think there's a reason to have both `isContainer()` and `isContainerEmpty()`, but since we do, technically the condition is `item.isRegularItem() || item.isFileAttachment()`. The `item.numAnnotations()` should be handled in `isContainerEmpty()`.

abaevbog replied:
Great catch, thank you! In fact, it already is handled in `isContainerEmpty`, so this must have been just a missed leftover from my initial attempt of getting annotations to show up in itemTree

abaevbog commented:
Bulk-tagging citations across multiple items (e.g. all citations from a collection) is another use case to keep in mind here: https://forums.zotero.org/discussion/113394/adding-tags-to-all-the-citations-in-a-collection

abaevbog commented:
To revisit this, I had it rebased onto the recent `fx115` and `CE` changes. Now one can select one or multiple annotations from the `itemTree` and they will be displayed in the `itemPane` in a single `collapsible-section` with a "Add Note from Annotations"/"Create Note from Annotations" button at the top. The title for each annotation in `itemTree` is constructed by adding the comment, annotation text and tags as long as there is space, and it seems quite descriptive to me.

Screenshot for reference:
<img width="1213" alt="Screenshot 2024-04-05 at 12 11 45 AM" src="https://github.com/zotero/zotero/assets/36271954/43c799ca-7006-4b14-b4d3-ef1beeb8060c">


What are other things that would be a good fit here? The only idea I had in mind is the "Move items to trash" option from the context menu. Would it make sense? I think there was some feedback in the past that if you accidentally delete an annotation, it's gone and cannot be returned. What if we just send annotation items into the trash like everything else, and the one would be able to erase or retrieve them if needed?

dstillman commented:
Neat! @yexingsha might have some thoughts here. 

Those vertical pipes aren't in the design, are they? I don't think we want those.

For the item pane, I'd think we might want headers for the parent item titles. Not sure we need much else.

We don't need a visible Move to Trash button — you should just be able to press Delete on the row like any other item — but that's a great point about the trash in general. We can totally send deleted annotations to the trash once this is in place.

larmarange commented:
This is great! 

A minor comment about the title of an annotation. For a text annotation, you used the user comment between quotes, followed by the highlighted text. For me, it would be more relevant to indicated in quotes the highlighted text (i.e. the verbatim extracted from the document), followed by the user comment.

dstillman commented:
> For me, it would be more relevant to indicated in quotes the highlighted text (i.e. the verbatim extracted from the document), followed by the user comment.

Yes, this is what it's supposed to be.

abaevbog commented:
> For the item pane, I'd think we might want headers for the parent item titles. Not sure we need much else.

Oh yes of course! The annotations should certainly be somewhat grouped by their parent item. The nicest way of doing it that I can think of is create a separate `collapsible-section` for each parent item and to use the name in the header. Maybe something like this?

<img width="364" alt="Screenshot 2024-04-05 at 9 17 50 AM" src="https://github.com/zotero/zotero/assets/36271954/5a8f6066-a093-4206-a6fc-3a7cd6f5ad18">

Alternatively, we could keep all annotations within the same `collapsible-section` like right now and insert a new mini-header component between `annotationRow`s. But I think it will look odd and we would loose the ability to collapse/expand sections. 

I tried to "shove" the displaying of multiple annotations into the existing `attachmentAnnotationBox` to reuse as much code as possible but now it looks like it might be better to create a separate component for it.

dstillman commented:
No, pipe being `|`, in the items-list row.

yexingsha commented:
This is great! I like the annotations being grouped by items in the item pane. My thoughts so far:

1. The tag dot is too close to the icon, should maintain the same amount of padding between text and icon.
2. In Figma I made annotation text and placeholders (image annotation, ink annotation) the same font size as tags (12px by default) to help distinguish elements. We can try to see how it feels? 
3. Do we need to display non-colored tags here? I feel like annotations should follow the same logic as items, so there's no need to display non-colored tags.
4. In the case of annotation text + comment, long annotation text can take up all the horizontal space, making it hard to tell if there is a comment. I think we should prioritize comment over annotation text, so in Figma I made comments take up as much space as possible, up to a maximum width (roughly 75%, but we can tinker with it).

abaevbog commented:
Circling back here after a bit of a break on accessibility-specific fixes.

@yexingsha this is great feedback, apologies for coming back to it after so long.

> The tag dot is too close to the icon, should maintain the same amount of padding between text and icon.

Fixed.

> In Figma I made annotation text and placeholders (image annotation, ink annotation) the same font size as tags (12px by default) to help distinguish elements. We can try to see how it feels?

Yup, I missed that part. I set the font to be the same as in tag-selector, which oddly was `font-size: 0.916666667em;` but it is a bit smaller now.

> Do we need to display non-colored tags here? I feel like annotations should follow the same logic as items, so there's no need to display non-colored tags.

That's fair. My thinking is that there is a max number of characters that the item title can have, so if there is a text and comment that are long enough, tags won't be included. But if the title is very short (e.g. "Image annotation"), I thought might as well add as much context? Or does it look too busy, in which case we can only add colored tags. Or no tags at all? 

> In the case of annotation text + comment, long annotation text can take up all the horizontal space, making it hard to tell if there is a comment. I think we should prioritize comment over annotation text, so in Figma I made comments take up as much space as possible, up to a maximum width (roughly 75%, but we can tinker with it).

There was part of the discussion [above](https://github.com/zotero/zotero/pull/3416#issuecomment-2039426672) where the consensus was to place the annotation text first with the comment after.  I tweaked how we put together a title to leave up to half of the available space for the comment, which makes sure that both will be visible if they are long. So now it looks kinda like this - what do you think?

<img width="1136" alt="Screenshot 2024-08-14 at 7 56 12 PM" src="https://github.com/user-attachments/assets/44b01192-e3a4-4555-86da-fd30754ddbb2">

There were a few additional tweaks to the sections in the `itemPane` and how they look - feel free to correct me if something doesn't look right there

Meanwhile, I'll iron out remaining broken tests.

yexingsha commented:
`font-size: 0.916666667em;` was from a previous iteration where default body text was 12px and tag text was 11px. Now that we've changed default body text to 13px, the correct number should be `font-size: 0.923076923em;`.

> But if the title is very short (e.g. "Image annotation"), I thought might as well add as much context? Or does it look too busy, in which case we can only add colored tags. Or no tags at all?

I think we should only display colored tags and emoji tags, which is consistent with items and attachments. In order to make text tags distinguishable from comments, we need to format them differently, which would add more visual complexity to an already busy row. Also, I think adding tags to some annotation but not all would cause some confusion.

Overall the current implementation of item tree looks good, some small issues:

<img width="919" alt="Screenshot 2024-08-14 at 22 50 53" src="https://github.com/user-attachments/assets/d2deec6f-11ec-46cf-a95e-ec4224ba6eca">

1. Emoji tags should have the same amount of padding as color dots; also need to remove the duplicate emoji at the end.
2. Annotations that are cut off should end in ellipsis.
3. Short comments that do not occupy over half of the width are not fully displayed.
4. I like that double clicking an annotation opens up the attachment, but it jumps to the top of the page where the annotation is located. Is it possible to jump to the actual position of the annotation and focus on it?

Issues in the item pane:

<img width="331" alt="Screenshot 2024-08-14 at 23 02 53" src="https://github.com/user-attachments/assets/d90f81ec-f8e0-4695-9c34-3ce29ee50f5b">

1. The background color of the header area should be `material-sidepane`, not `material-toolbar`.
2. "Create Note from Annotations" is not vertically centered in the header area. This is because macOS native elements in fx115 is weird, which should be fixed with [these modifications](https://github.com/zotero/zotero/pull/3807#issuecomment-2240539160). I'm not sure why they aren't applied here. 
3. Let's cut off item titles by letter instead of words to display the maximum amount of information possible.
5. There is too little gap between annotations, which should be 6px in compact and 8px in comfortable.
6. In long comments being cut off, (what I assume is) the next line is peeking through the border. This issue also appears in beta, but doesn't appear in windows. It's also more noticeable in dark mode, which I guess is why I haven't noticed it before.

abaevbog commented:
> font-size: 0.916666667em; was from a previous iteration where default body text was 12px and tag text was 11px. Now that we've changed default body text to 13px, the correct number should be font-size: 0.923076923em;.

Cool, I changed the font of both the tag selector and the annotation rows in the table for `0.923076923em`

> I think we should only display colored tags and emoji tags, which is consistent with items and attachments. In order to make text tags distinguishable from comments, we need to format them differently, which would add more visual complexity to an already busy row. Also, I think adding tags to some annotation but not all would cause some confusion.

Fair enough, I dropped the addition of tags into the display name of the annotation. That way, everything across the table rows is consistent.

> Emoji tags should have the same amount of padding as color dots; also need to remove the duplicate emoji at the end.

Added missing margins, should be fixed. Now that we don't try to plug tags into the name, they're also gone from the end.

> Annotations that are cut off should end in ellipsis.

Gotcha, we'll add `Zotero.getString("punctuation.ellipsis")` when annotation text or comment is cutoff.

> Short comments that do not occupy over half of the width are not fully displayed.

Thanks for pointing that out - it was a glitch due to not accounting for added `"` around the title while calculating allowed comment length. Should be fixed now.

> The background color of the header area should be material-sidepane, not material-toolbar.

Done

> "Create Note from Annotations" is not vertically centered in the header area. 

I copied the stylesheet from `_itemMessagePane.scss` and it had `margin: 0` overriding some of those css tweaks from. I removed it, so it should look better now?

> Let's cut off item titles by letter instead of words to display the maximum amount of information possible.

Done

>  There is too little gap between annotations, which should be 6px in compact and 8px in comfortable.

Right. I assume it should be the same as in [attachmentAnnotationBox.scss](https://github.com/zotero/zotero/blob/596703b21b892d2a673b0103bac9077e82f75942/scss/elements/_attachmentAnnotationsBox.scss#L13)? It has 4px in compact mode, not 6px, though. Which one is correct?

> In long comments being cut off, (what I assume is) the next line is peeking through the border. This issue also appears in beta, but doesn't appear in windows. It's also more noticeable in dark mode, which I guess is why I haven't noticed it before.

Should be fixed. I wonder, though, if we need to cutoff the annotations at 9 lines when an annotation row is selected? Presumably, if you are clicking on it, you are specifically looking at annotations, so maybe they should appear fully expanded? I tried it here, let me know what you think.

This is what it's looking like now:

<img width="1261" alt="Screenshot 2024-08-15 at 12 30 46 PM" src="https://github.com/user-attachments/assets/50e644b3-b318-47ed-8a91-f43092f414af">

abaevbog commented:
> I like that double clicking an annotation opens up the attachment, but it jumps to the top of the page where the annotation is located. Is it possible to jump to the actual position of the annotation and focus on it?

This last point is actually a good question. @mrtcode, on selection of an annotation row [here](https://github.com/zotero/zotero/blob/107558d60eb50d8a05d3c00a7564d11376d730e2/chrome/content/zotero/zoteroPane.js#L4770) I am calling `ZoteroPane.viewPDF` in a similar manner as it is done in [editorInstance.js](https://github.com/zotero/zotero/blame/596703b21b892d2a673b0103bac9077e82f75942/chrome/content/zotero/xpcom/editorInstance.js#L496). It works as expected sometimes. Some other times, though, it just opens up the reader without scrolling to the annotation. Am I missing something else? For reference: [video](https://www.dropbox.com/scl/fi/zbga30vxwzpm9otmpcy6u/viewPDF_trimmed.mov?rlkey=fxw9bpq5c04eh4d9cksv89wxn&st=pr8q7jmc&dl=0).
 
I think it's somewhat similar to what happens when I open the annotation from the note-editor though. Generally, it scrolls to the correct position but sometimes it just open up the reader.

yexingsha commented:
Short comments are still not fully displayed:

<img width="925" alt="Screenshot 2024-08-15 at 15 26 18" src="https://github.com/user-attachments/assets/b0edee1f-e5c6-432e-9db6-b31cc6b46bed">

> I copied the stylesheet from `_itemMessagePane.scss` and it had `margin: 0` overriding some of those css tweaks from. I removed it, so it should look better now?

The horizontal margins are fixed, but the vertical margins are still weird. Seems like `@media (-moz-platform: macos) {margin: 1px -2px -1px;}` does the trick. It really irks me that we have to manually adjust this for every other button, but so be it.

> I assume it should be the same as in [attachmentAnnotationBox.scss](https://github.com/zotero/zotero/blob/596703b21b892d2a673b0103bac9077e82f75942/scss/elements/_attachmentAnnotationsBox.scss#L13)? It has 4px in compact mode, not 6px, though. Which one is correct?

Apologies, should be 4px in compact mode.

> I wonder, though, if we need to cutoff the annotations at 9 lines when an annotation row is selected? Presumably, if you are clicking on it, you are specifically looking at annotations, so maybe they should appear fully expanded? I tried it here, let me know what you think.

I think that makes sense. As long as no one is creating annotations that can take up the entire height of the item pane (and I assume most people aren't), I don't see any downside to this.

abaevbog commented:
> Short comments are still not fully displayed:

Hold on, I may have misunderstood the desired behavior here... I was under the impression that we should construct the annotation name as one string `"${annotation_text}" ${annotation_comment}` and place it as one cell into the row of table. Not sure where that came from, I suppose I thought so because if it's not one single string, it is like having columns that don't correspond to headers at the top and I thought it would feel weird. That means that, depending on how wide the itemPane is expanded, at some point the comment part of the string will get cutoff. 

https://github.com/user-attachments/assets/01ebcd02-9e45-46c4-b649-fab0968c852f

Did I misunderstand it? Do we want the annotation text and annotation comment to be treated sorta like columns, where annotation text always occupies, say, 2/3 of the width, and the comment occupies the rest? In that case, if the comment fits into that 1/3 of the space, it would display fully, and when the table shrinks, both columns would shrink by a little bit, instead of the comment being hidden first. Something like this mockup?


https://github.com/user-attachments/assets/580c6d01-38ea-4fd9-9e1d-485f3099038a

abaevbog commented:
> The horizontal margins are fixed, but the vertical margins are still weird. 

How about now? `align-items: center` on the `.custom-head` seems to do it without hard-coding specific margins. I added it to the `_itemMessagePane.scss` as well since the "Restore to Library" and "Delete Permanently" buttons when trash is selected looked not aligned as well.

yexingsha commented:
Sorry for not explaining the desired behavior clearly. My thinking is that we should always show some comment even at smaller widths, since whether an annotation contains comment is an important distinction, and comments are arguably more important than the annotation texts. So yes, the second mockup is much closer to the desired behavior. I would say annotation text and comment each takes up roughly 1/2 of the width, with comment getting a bit more priority.

The vertical alignment in the header looks fine now, but we still need to replace the `.custom-head` button margin with `@media (-moz-platform: macos) {margin: 0px -2px;}` for the horizontal margins to look right.

mrtcode commented:
> > I like that double clicking an annotation opens up the attachment, but it jumps to the top of the page where the annotation is located. Is it possible to jump to the actual position of the annotation and focus on it?
> 
> This last point is actually a good question. @mrtcode, on selection of an annotation row [here](https://github.com/zotero/zotero/blob/107558d60eb50d8a05d3c00a7564d11376d730e2/chrome/content/zotero/zoteroPane.js#L4770) I am calling `ZoteroPane.viewPDF` in a similar manner as it is done in [editorInstance.js](https://github.com/zotero/zotero/blame/596703b21b892d2a673b0103bac9077e82f75942/chrome/content/zotero/xpcom/editorInstance.js#L496). It works as expected sometimes. Some other times, though, it just opens up the reader without scrolling to the annotation. Am I missing something else? For reference: [video](https://www.dropbox.com/scl/fi/zbga30vxwzpm9otmpcy6u/viewPDF_trimmed.mov?rlkey=fxw9bpq5c04eh4d9cksv89wxn&st=pr8q7jmc&dl=0).
> 
> I think it's somewhat similar to what happens when I open the annotation from the note-editor though. Generally, it scrolls to the correct position but sometimes it just open up the reader.

It needs a location parameter, which can have position, pageIndex, pageLabel, or annotationID:

```
this.viewPDF(item.parentItemID, { position: JSON.parse(item.annotationPosition)});
```

Or maybe it would be better to pass the annotationID instead of the position. The only reason we currently pass the position from the note editor is because the annotation might no longer exist, which wouldn’t be an issue in this case.

dstillman commented:
For my future reference:

https://forums.zotero.org/discussion/114531/zotero-7-search-for-tags-in-annotations
https://forums.zotero.org/discussion/116547/newbie-question-aggregating-quotes-from-multiples-sources-by-tag

(lots other earlier ones)

yexingsha commented:
Some more ideas for how to differentiate annotation text and comment:

1. Treat text and comment like columns and add more spacing between them:
	<img width="661" alt="Screenshot 2024-08-16 at 14 18 42" src="https://github.com/user-attachments/assets/26077631-62e8-40fe-8c0c-4e123572ef10">


2. Annotation text in 12px, comments and titles in 13px:	
	<img width="661" alt="Screenshot 2024-08-16 at 14 01 40" src="https://github.com/user-attachments/assets/d09cb190-4c95-4d65-9f61-f1361925fa32">	

3. Underline annotation text:
	<img width="661" alt="Screenshot 2024-08-16 at 14 22 19" src="https://github.com/user-attachments/assets/4a161409-8f6c-4f73-801d-c31a83e0e04a">


4. I can't recommend italics because it looks horrible for CJK characters:
	<img width="661" alt="Screenshot 2024-08-16 at 14 21 13" src="https://github.com/user-attachments/assets/ed4a0156-8ba3-4a79-9e6b-556aa6470079">

abaevbog commented:
@yexingsha I refactored how we put together the row for the table, so now the title and comment are essentially not much different from other columns of the table in terms of sizing: a bit of both comment and title are always visible, if the comment is short, it will be fully displayed for some time but if the table gets narrower, it will be cutoff at some point. This is the behavior right now:


https://github.com/user-attachments/assets/eec37c83-bac0-4b3a-a9cf-98da8d6a87ab

I agree that comments and quoted text needed to be more differentiated (though adding proper quotes helped a lot). For now, I went with your suggested option 1 where I kept the column-style spacing between them. They essentially are like columns (just without headers), so it felt appropriate to me. Let me know what you think about it now, we can try different fonts and such as well. Personally, I find the underline from option 3 too busy, and my favorite is actually italics from option 4. I think it makes it immediately clear that these are different components without sacrificing readability. Also it feels appropriate to italicize a direct quote. It's a shame it doesn't work well with non-latin characters... 

Finally, I added the `@media (-moz-platform: macos) {margin: 0px -2px;}` margins to header buttons. Is it looking better? Apologies for this margin-related silliness, somehow I always struggle with issues of sizing and positioning and at times do not see if things are aligned or not.

abaevbog commented:
@mrtcode thank you for the explanation! Per your advice, I did `this.viewPDF(item.parentItemID, { annotationID: item.key });` and it works great.

dstillman commented:
Yeah, italics really does work the best for Latin characters. We could disable italics when there are CJK characters and add spacing in that case?

abaevbog commented:
Yeah, let's try that!

yexingsha commented:
The header margins look perfect now!

And yeah let's try using italics for non-CJK characters. There are definitely other scripts that should not use italics, but a cursory search didn't turn up a convenient list for those, so I guess we will add more exceptions to this rule when people bring it up.

The only issue I see right now is that comments doesn't seem to consistently take up enough width. In the second selected annotation in this screenshot, the comment is as long as the annotation text, but is using very little width:

<img width="1582" alt="Screenshot 2024-08-16 at 17 37 06" src="https://github.com/user-attachments/assets/40bff067-d930-45bf-a70e-57e07ee80e25">

abaevbog commented:
> And yeah let's try using italics for non-CJK characters. There are definitely other scripts that should not use italics, but a cursory search didn't turn up a convenient list for those, so I guess we will add more exceptions to this rule when people bring it up.

Cool, so I added a `_containsCJKCharacters` function to check against a number of regular expressions that, for now, look for characters from the following unicode ranges: CJK Radicals Supplement,  CJK Symbols and Punctuation, Hiragana (Japanese), Katakana (Japanese), Hangul Jamo (Korean), Enclosed CJK Letters and Months, CJK Compatibility, CJK Unified Ideographs Extension A, CJK Unified Ideographs, Hangul Syllables (Korean), CJK Compatibility Ideographs,  CJK Compatibility Forms, / CJK Unified Ideographs Extension B, CJK Compatibility Ideographs Supplement. Let me know if I should add/remove something from here. We can also extend the list later based on feedback. 

Now, if there is a CJK character in the quote, we leave default spacing for columns and remove italics:

<img width="1176" alt="Screenshot 2024-08-18 at 3 40 19 PM" src="https://github.com/user-attachments/assets/9d899575-f73f-4c42-b2eb-8999f1b50401">

If there is no CJK character in the quote, spacing is a bit smaller and the quote is in italics:
<img width="788" alt="Screenshot 2024-08-18 at 3 40 38 PM" src="https://github.com/user-attachments/assets/af7af4c5-5b91-43d1-a51a-0c985f967b0a">


> The only issue I see right now is that comments doesn't seem to consistently take up enough width. In the second selected annotation in this screenshot, the comment is as long as the annotation text, but is using very little width:

I had kept the behavior where the first part of the annotation would be counted as part of the "title" column so when the actual title column is widened/shrunk, it would resize with it. So if the title column was wider, the quote part of the annotation would be wider and the comments would occupy relatively less space.

https://github.com/user-attachments/assets/1fdc9c22-2d36-4dfd-b284-336f16598ba7

If this is not something we want, I removed it and the ratio of comment to quote should now always be roughly 2:1. This is the current behavior:

https://github.com/user-attachments/assets/d972a281-1d69-4b26-84d3-c106d1741b32

Let me know what you think!

dstillman commented:
> Now, if there is a CJK character in the quote, we leave default spacing for columns and remove italics

What would happen if we still italicized non-CJK text within a row, and only added the extra space if the visible text ended with a CJK character?

Re: column resizing, I would say definitely the second one. They're both weird, but the first one seems way weirder to me — this isn't a title, not all annotations actually have this text (so only some would move), and it's not even aligned with the column.

yexingsha commented:
The unicode ranges look fine to me. Also agree with @dstillman on both counts.

abaevbog commented:
Cool! I think italicizing the non-cjk characters makes a lot of sense. 

> only added the extra space if the visible text ended with a CJK character?

The only complexity with this is that the visible characters are still not static because the annotation's quote and comment will shrink/expand when the itemPane is expanded/shrunk. That means that we need to keep track of when the last visible character changes due to itemPane resizing (also font changes) and add/remove spacing dynamically. Wouldn't that be a bit strange and jumpy?

For reference, this is what I mean by the last visible character being not static. In the first screenshot, the `互` symbol is not visible, so we would not want any spacing but then, once I make the itemPane narrower, it appears, and the comment would need to be "pushed" away from the quote when that happens.

<img width="1145" alt="Screenshot 2024-08-19 at 11 24 22 AM" src="https://github.com/user-attachments/assets/269f3f4f-cb20-441c-858a-9c78c48f4b31">
<img width="1133" alt="Screenshot 2024-08-19 at 11 24 09 AM" src="https://github.com/user-attachments/assets/83e40502-7576-47cb-a572-b6c54f42834e">

yexingsha commented:
That makes sense. Perhaps it's better to always keep the extra space for the increased legibility?

abaevbog commented:
That's what I am leaning to... We could still remove spacing if there are no CJK characters at all though, since with just latin characters we do not have this problem

dstillman commented:
Yeah, let's try only adding a space if there are any CJK characters.

abaevbog commented:
Cool! What do we think about this?

- non-CJK characters from the quote are wrapped in italics `<i>`
- CJK characters are left as is
- if there is a CJK character in the quote, use default horizontal padding of 8px from the table (first 2 rows). Otherwise use smaller padding of 2px to keep the quote closer to the comment 

<img width="1173" alt="Screenshot 2024-08-19 at 7 54 12 PM" src="https://github.com/user-attachments/assets/9a17892d-5c45-407c-9e34-4b29de90fe09">

yexingsha commented:
Looks good to me!

abaevbog commented:
Great! I rebased to resolve the conflicts, clean up some of the intermediary changes, and fix up a few smaller glitches I ran into.

@adomasven we ended up rendering annotation rows differently from when you reviewed last. To have comment and quote always be visible, they are now rendered somewhat like two columns (but without corresponding to actual columns from the table). I ended up treating annotation rows as a completely separate case in `_renderItem` of `itemTree.jsx` where we decide how each annotation is rendered. Correct me if there is a better place for this logic.

abaevbog commented on chrome/content/zotero/elements/annotationItemsPane.js:
```diff
@@ -0,0 +1,136 @@
+						`<collapsible-section
+							data-l10n-id="section-attachments-annotations"
+							data-pane="annotations-${parentItem.id}"
+							summary="${parentItem.getDisplayTitle()}"
```
Ah, great catch! Thanks!

abaevbog commented on chrome/content/zotero/elements/annotationItemsPane.js:
```diff
@@ -0,0 +1,136 @@
+				// Create a collapsible section for each top-level item if it does not exist yet
+				let section = this.querySelector(`[data-pane="annotations-${parentItem.id}"]`);
+				if (!section) {
+					section = MozXULElement.parseXULToFragment(
+						`<collapsible-section
+							data-l10n-id="section-attachments-annotations"
+							data-pane="annotations-${parentItem.id}"
+							summary="${parentItem.getDisplayTitle()}"
+							data-item-id="${parentItem.id}">
+
+							<html:div class="body"></html:div>
+
+						</collapsible-section>`
+					).querySelector("collapsible-section");
```
Good to know, I'll keep it in mind for the future! The only reason I went with this is for slightly better readability. There was no functional reason not to just `document.createXulElement` the `collapsible-section`, add a `.body` child and set the attributes.

abaevbog commented on chrome/content/zotero/itemTree.jsx:
```diff
@@ -3903,6 +3998,76 @@ var ItemTree = class ItemTree extends LibraryTree {
+	// Check if text quoted in an annotation row contains CJK characters to determine
+	// if it should be italicized or not
+	_containsCJKCharacters = (str) => {
+		let cjkRegexArray = [
+			new RegExp("[\u2E80-\u2EFF]"), // CJK Radicals Supplement
+			new RegExp("[\u3000-\u303F]"), // CJK Symbols and Punctuation
+			new RegExp("[\u3040-\u309F]"), // Hiragana (Japanese)
+			new RegExp("[\u30A0-\u30FF]"), // Katakana (Japanese)
+			new RegExp("[\u1100-\u11FF]"), // Hangul Jamo (Korean)
+			new RegExp("[\u3200-\u32FF]"), // Enclosed CJK Letters and Months
+			new RegExp("[\u3300-\u33FF]"), // CJK Compatibility
+			new RegExp("[\u3400-\u4DBF]"), // CJK Unified Ideographs Extension A
+			new RegExp("[\u4E00-\u9FFF]"), // CJK Unified Ideographs
+			new RegExp("[\uAC00-\uD7AF]"), // Hangul Syllables (Korean)
+			new RegExp("[\uF900-\uFAFF]"), // CJK Compatibility Ideographs
+			new RegExp("[\uFE30-\uFE4F]"), // CJK Compatibility Forms
+			new RegExp("[\u{20000}-\u{2A6DF}]", "u"), // CJK Unified Ideographs Extension B
+			new RegExp("[\u{2F800}-\u{2FA1F}]", "u") // CJK Compatibility Ideographs Supplement
+		];
+		return cjkRegexArray.some(exp => exp.test(str));
```
`u` was required on the last two because without it `Range out of order in character class` error is thrown. It didn't seem necessary for other expressions...

And the only reason I did it as an array is for readability. This way, it's easier to see which regex refers to which range of characters, which I think might be helpful in the future if we ever need to go back to edit this to include/exclude some more ranges. With primitive syntax, I don't believe we can break the expression into multiple lines to place comments next to each range, can we? In that case, we would have something like this which to me felt a bit harder to work with:
```
// CJK Radicals Supplement: \u2E80-\u2EFF
// CJK Symbols and Punctuation: \u3000-\u303F
// Hiragana (Japanese): \u3040-\u309F
// Katakana (Japanese): \u30A0-\u30FF
// ... names of remaining ranges 
let cjkRegex = /[\u2E80-\u2EFF\u3000-\u303F\u3040-\u309F\u30A0-\u30FF\u1100-\u11FF\u3200-\u32FF\u3300-\u33FF\u3400-\u4DBF\u4E00-\u9FFF\uAC00-\uD7AF\uF900-\uFAFF\uFE30-\uFE4F\u{20000}-\u{2A6DF}\u{2F800}-\u{2FA1F}]/u;
```

In any case, there was never any need to have multiple `RegExp`. It could be a single `RegExp` with a large concatenated string inside. Would that be better? Or do you think it's worth to make it one long primitive for performance reasons?

abaevbog replied:
Oh, sure! The `Script=Han` would not include all the Japanese and Korean ranges but I think `/[\p{Script=Han}|\p{Script=Hiragana}|\p{Script=Katakana}|\p{Script=Hangul}]/u` should do the trick and is, indeed, pretty self explanatory.

abaevbog commented on chrome/content/zotero/itemTree.jsx:
```diff
@@ -3903,6 +3998,76 @@ var ItemTree = class ItemTree extends LibraryTree {
+	 * @returns {Node[]} resultArr - An array of nodes preserving the original order, with non-CJK text segments italicized.
+	 * If the input node is not a text node, new segments are set as children of the input node, d
+	 * and the returned resultArr contains only the original node.
+	 */
```
Yes, we can ... Very neat! I didn't know about `font-synthesis` - fonts turn out to be way smarter than i thought.

abaevbog commented on chrome/content/zotero/elements/annotationRow.js:
```diff
@@ -96,7 +96,7 @@
 				let text = document.createElement('div');
 				text.classList.add('quote');
-				text.textContent = this._annotation.annotationText;
+				text.innerHTML = this._annotation.annotationText;
```
That's fair, I didn't think about it. I stumbled upon this [mozilla util](https://searchfox.org/mozilla-central/source/parser/html/nsIParserUtils.idl#100) that might be helpful here. From what I can tell, it strips attributes and non-style tags from components within a string.

So if we do:

```
let parserUtils = Cc["@mozilla.org/parserutils;1"].getService(Ci.nsIParserUtils);
parserUtils.sanitize("<div attribute='something'>this</div> <unknown>is</unknown> a <b>test</b><script>console.log('!')</script> <img src='test'>", Ci.nsIParserUtils.SanitizerDropForms | Ci.nsIParserUtils.SanitizerDropMedia)
```

We'll get `<html><head></head><body><div>this</div> is  a <b>test</b> </body></html>`. I pushed a commit with that. What do you think? Or, if it's better, I can revert it back to textcontent for now and we can tackle it separately.

abaevbog replied:
It looks like itemTree already does essentially [the same thing as the reader](https://github.com/zotero/zotero/blob/f8970b92de368a0ddb8887a1ba7dfa3c0548bdd4/chrome/content/zotero/itemTree.jsx#L2752) to properly render the title column. Now it feels even less nice to duplicate reader's logic... Would it make sense to pull out this code from `itemTree` into, maybe, `Zotero.Utilities.Internal` so that `itemTree` could share it with `annotation-row`?

abaevbog replied:
Ah, yes, I see it. Then the current rendering of annotation titles as a primary cell of the table is not right as well. Then I suppose the best way is indeed to just copy over the logic from the reader (into Zotero.Utilities.Internal as well?) and use it for annotation rows of the itemTree and the `annotation-row` component.

abaevbog replied:
That makes a lot of sense! Trying to do it now is almost a guarantee to create conflicts. I pushed a tweak to strip all html tags for now and then, once these PRs are merged, I can move over the rendering logic from the reader into that new utilities function as another mode.

abaevbog commented on chrome/content/zotero/elements/annotationRow.js:
```diff
@@ -96,7 +96,18 @@
+				}
+				else {
+					text.textContent = this._annotation.annotationText;
+				}
```
Sorry, I'm a bit confused... After our discussion [last week](https://github.com/zotero/zotero/pull/3416#discussion_r1727941691), I pushed https://github.com/zotero/zotero/pull/3416/commits/4bc0393530a1bdf21712f20139906f9af5a36597 to just strip all html tags and render everything as plain text. That is, until all relevant PRs such as https://github.com/zotero/zotero/pull/4607 are merged to avoid conflicts. So all that should be happening now is a [call](https://github.com/zotero/zotero/blob/e28002bd2c7ef72a5df4b13de4ee8b0850c1a81f/chrome/content/zotero/elements/annotationRow.js#L103) to `convertToPlainText`.

This is what this part looks like this to me on this branch:
```
if (this._annotation.annotationText) {
  let text = document.createElement('div');
  text.classList.add('quote');
  let plainQuote = parserUtils.convertToPlainText(this._annotation.annotationText, Ci.nsIDocumentEncoder.OutputRaw, 0);
  text.textContent = plainQuote;
  this._body.append(text);
}
``` 

Is that not what is appearing on your end?

abaevbog replied:
Gotcha! No worries!

abaevbog commented on chrome/content/zotero/itemTree.jsx:
```diff
@@ -1194,7 +1195,7 @@ var ItemTree = class ItemTree extends LibraryTree {
 			rowsToSelect.push(row);
 		}
-		
+		console.log("Rows to select ", rowsToSelect)
```
Ah, yes. Thanks!

dstillman commented:
When double-clicking on an annotation, it opens the reader but doesn't flash it the way that we (sometimes?) do when focusing it from notes. @mrtcode, is there a way to trigger that?

dstillman commented:
Currently, including an annotation in a multi-item selection doesn't affect the available options — e.g., if you select a regular item and an annotation, the regular-item options (e.g., "Export Items…") are still available. Same for an attachment and its annotations.

Should including an annotation limit the context menu options to just the annotation options ("Add Note from Annotations" or "Create Note from Annotations"), based on the selected annotations and the descendant annotations of any selected items?

We do show "Create Note from Annotations" now if an item with descendant annotations and an item without descendant annotations are selected, so there's precedent for ignoring irrelevant items for that option. Whereas when, say, "Export Items…" appears, it's because it applies to the full selection, which is no longer the case in this PR.

mrtcode commented:
> When double-clicking on an annotation, it opens the reader but doesn't flash it the way that we (sometimes?) do when focusing it from notes. @mrtcode, is there a way to trigger that?

Yes, that needs to be addressed in the Reader. Currently, annotation ID doesn't trigger flashing — only navigation to a specific position does.

#5211

dstillman commented:
For an ink or image annotation that's not available, we should probably do something better than a broken image:

![Screenshot 2025-04-16 at 7 38 16 AM](https://github.com/user-attachments/assets/44c297cf-0861-401b-beb0-9183660da65e)

Not sure what. Just "[Image not available]"?

abaevbog commented:
> Should including an annotation limit the context menu options to just the annotation options ("Add Note from Annotations" or "Create Note from Annotations"), based on the selected annotations and the descendant annotations of any selected items?

I was under the impression that the context menu options are meant to appear if they are appropriate for ANY of the selected items (not ALL). For example, if a PDF attachment and a note are selected, I still get the "Open PDF in New Tab" option that's not relevant for the note). But I do agree that it doesn't make much sense to keep all those options open. So the general logic now is to keep only annotation-related options (create or add annotation) when any of the selected rows are annotations.

> For an ink or image annotation that's not available, we should probably do something better than a broken image:

Yeah, I think "[Image not available]" seems appropriate. This is what it would look like now. What do you think?

<img width="326" alt="Screenshot 2025-04-16 at 10 36 47 PM" src="https://github.com/user-attachments/assets/e57cea2b-0205-482f-9266-2941e63e1abd" />

abaevbog commented on chrome/content/zotero/elements/annotationRow.js:
```diff
@@ -84,24 +84,40 @@
-					this._body.append(img);
+					// If the file actually exists, display it. Otherwise, show a placeholder.
+					let file = Zotero.File.pathToFile(imagePath);
+					if (file.exists()) {
```
That's a good idea! There is a brief flash if the file is missing between when the `<img>` node is inserted and when the `error` event fires which sets the placeholder text. But it's not too bad and I agree it's worth it.

abaevbog replied:
This is what it's looking like now. Do you think it's noticeable enough to be worth it?

https://github.com/user-attachments/assets/f69a892d-e482-40aa-add4-604a4ea0e966

abaevbog replied:
Cool! Yes, it really should be quite rare. If the annotation exists but the actual `.png` file is maybe deleted by the user, or something like that.

dstillman replied:
That looks fine to me, but this won't be rare at all — it would happen any time data has synced but not the PDF. That would even include when people have files set to download "as needed" and they haven't opened the file on this computer yet.

abaevbog commented:
During the last edits, I ran into a few additional questions to discuss:
- when all selected items are being expanded by pressing right arrow, should attachment rows be expanded to display annotation rows? Currently that does not happen because I thought it would visually not be good to (potentially) add so many additional rows in one go. But on the other hand maybe it would be more useful?
- I realized that we want to support quick copy for the annotation rows in some form, right? Currently, dragging an annotation row will attempt to generate bibliography in `Zotero.QuickCopy.getContentFromItems` via `Zotero.Cite.makeFormattedBibliographyOrCitationList`. It eventually fails in [itemToCSLJSON](https://github.com/zotero/utilities/blob/f9c86c0d62b492b5b703c9b442f52610979df8fd/utilities_item.js#L66) because annotation items do not have a cslType. My thinking now is to replicate how we generate the content of a note when it is created from annotations, and maybe use that as `text/html` data. Or we could actually make a note object (without saving though) and pass that into `Zotero.QuickCopy.getContentFromItems`. I'll experiment with these approaches but let me know if there is something else to consider that I am missing!
https://github.com/zotero/zotero/blob/a7778b93e841ce62773efc65cd86576c7bfe8af1/chrome/content/zotero/xpcom/editorInstance.js#L1391-L1410

dstillman commented:
> when all selected items are being expanded by pressing right arrow, should attachment rows be expanded to display annotation rows?

No, only selected rows should be expanded. The current behavior is correct.

You would use + to expand all items, including all descendant items.

> Or we could actually make a note object (without saving though) and pass that into Zotero.QuickCopy.getContentFromItems

Yes, we do exactly this in the reader.

https://github.com/zotero/zotero/blob/a7778b93e841ce62773efc65cd86576c7bfe8af1/chrome/content/zotero/xpcom/reader.js#L358-L426

We probably want to pull most of this into a separate method in `Zotero.QuickCopy`, since we want the exact same behavior here. (@mrtcode, is that reasonable?)

abaevbog commented:
Thank you for the explanation about the annotation/notes workflow!

I pulled out the logic of creating a temp note object with serialized annotations into `Zotero.QuickCopy.wrapAnnotationsAsNote`. One can now either drag-drop annotations into an editor or copy them via Edit > Copy Annotations menu.

One thing to note here is that we need to convert `Zotero.Item` annotations to JSON. `ink` and `image` annotations require fetching the file, which means `Zotero.Annotation.JSON` is `async`. `Zotero.QuickCopy.wrapAnnotationsAsNote` has to be synchronous because it's used in synchronous functions. So I pulled out the synchronous part of file `Zotero.Annotation.JSON` into  `Zotero.Annotation.JSONsync`, and just skip `ink` and `image` annotations. This felt appropriate because they don't work with drag-drop export in reader annotations anyways. If I drag an `ink` annotation from the reader, nothing happens. If I drag `image` annotation, I end up with "Error! Filename not specified." error in the text editor. Does this make sense?

dstillman commented on chrome/content/zotero/xpcom/quickCopy.js:
```diff
@@ -308,7 +308,44 @@ Zotero.QuickCopy = new function() {
+	this.wrapAnnotationsAsNote = function (annotations) {
+		let jsonAnnotations = [];
+		for (let annotation of annotations) {
+			if (annotation instanceof Zotero.Item) {
```
When would these not be `Zotero.Item` objects? Can we not just require that?

abaevbog replied:
In reader `onSetDataTransferAnnotations`, the passed annotations are already converted to JSON, so we just use those objects right away. But in case of drag-drop of annotation rows or "Copy Annotations" menu option, I thought it's better to be able to pass the actual `Zotero.Item`, so we don't  have to also call `Zotero.Annotations.toJSONSync` and set `attachmentItemID` before each `wrapAnnotationsAsNote`.

dstillman commented on chrome/content/zotero/xpcom/quickCopy.js:
```diff
@@ -308,7 +308,44 @@ Zotero.QuickCopy = new function() {
+	 * If an array of Zotero.Items is passed, they will be converted to JSON.
+	 * @returns {Zotero.Item} - A note item with the annotations serialized as HTML.
+	 */
+	this.wrapAnnotationsAsNote = function (annotations) {
```
`annotationsToNote`?

abaevbog replied:
Yeah, `annotationsToNote` sounds better

dstillman commented on chrome/locale/en-US/zotero/zotero.ftl:
```diff
@@ -96,6 +96,9 @@ menu-view-columns-move-left =
     .label = Move Column Right
 
+menu-edit-copy-annotation =
+    .label = Copy Annotation
```
When multiple annotations are selected, this should say "Copy [n] Annotations", using `$count`.

dstillman replied:
(We don't actually do this properly for "Copy Citation", but we might as well get this one right.)

dstillman replied:
(Or it could be "Copy Annotations", but that's probably not any easier. Finder does "Copy 5 Items" for multiple-selection.)

abaevbog replied:
Totally! The "Copy Citation" is actually what threw me off here - I thought maybe the annotations menu should be consistent with it. But you're right, might as well make this one right. So it says "Copy [n] Annotations" for many annotation rows now.

dstillman commented:
Oh, we can't do this:

![Screenshot 2025-04-25 at 1 52 20 AM](https://github.com/user-attachments/assets/7b2b1af5-5712-437e-bd28-f7f05b131702)

I assume this is coming from here:

https://github.com/zotero/zotero/pull/3416/files#diff-25ca00cc00683daf35b34ab7f21d87a19fb013badf4995f73bd810df83610600R92

https://github.com/zotero/zotero/blob/3144314d20e51fb8cfb73d65548122c9fe1eb9ed/chrome/content/zotero/elements/collapsibleSection.js#L349

We'll need to somehow avoid that. There should be a single state for the Annotations section in the attachment box (`extensions.zotero.panes.attachment-annotations.open`, I think), but then I don't think the Annotations section when annotations are selected even needs to be collapsible at all, at least unless we start showing anything else in that pane.

abaevbog commented:
Oooh, yeah, that is definitely not right. Thank you, I'll see how to avoid this

dstillman commented:
It looks like we can easily disable both collapsing and saving of the open state in collapsibleSection.js.

dstillman commented:
(Good work @windingwind!)

dstillman commented:
@windingwind: Instead of hard-coding ids, can we just set `.collapsible = false` when creating the section?

That seems to stop `open()` from doing anything, but it looks like the twisty is only hidden based on `_disableCollapsing()`. Is `.collapsible` just for empty sections with no content?

It would be nice to be able to set the equivalent of `_disableCollapsing()` without having to hard-code ids.

(And is it necessary to set both, even now? If it's always open, will it ever try to save an open state? It seems like one should be enough.)

dstillman commented:
@abaevbog: Anything else here from you, or are we good to merge?
